README.md
# MetaStreet Contracts

## Prerequisites

Node v16 or below. Hardhat doesn't work with Node v17 currently: [hardhat#1988](https://github.com/nomiclabs/hardhat/issues/1988).

## Usage

Install:

```
npm install
```

Compile contracts:

```
npm run build
```

Run unit tests:

```
npm test
```

Start hardhat network:

```
npm run node
```

## Additional Targets

- Format contracts (prettier): `npm run format`
- Lint contracts (solhint): `npm run lint`
- Run unit tests with gas reporter: `npm run test:gas`
- Run unit tests with coverage (solidity-coverage): `npm run test:coverage`
- Run static analyzer (slither, requires external installation): `npm run analyze`
- Format tests and scripts (prettier): `npm run format:ts`
- Lint tests and scripts (eslint): `npm run lint:ts`

## File Structure

- [`contracts/`](contracts/) - Smart Contracts
  - [`interfaces/`](contracts/interfaces) - Interfaces
    - [`INoteAdapter.sol`](contracts/interfaces/INoteAdapter.sol) - NoteAdapter interface
    - [`ILoanPriceOracle.sol`](contracts/interfaces/ILoanPriceOracle.sol) - LoanPriceOracle interface
    - [`ILoanReceiver.sol`](contracts/interfaces/ILoanReceiver.sol) - LoanReceiver interface
    - [`IVault.sol`](contracts/interfaces/IVault.sol) - Vault interface
    - [`IVaultRegistry.sol`](contracts/interfaces/IVaultRegistry.sol) - VaultRegistry interface
  - [`LoanPriceOracle.sol`](contracts/LoanPriceOracle.sol) - LoanPriceOracle implementation
  - [`LPToken.sol`](contracts/LPToken.sol) - LPToken implementation
  - [`Vault.sol`](contracts/Vault.sol) - Vault implementation
  - [`VaultRegistry.sol`](contracts/VaultRegistry.sol) - VaultRegistry implementation
  - [`integrations/`](integrations/) - Thirdparty integration contracts
    - [`NFTfiV2/`](contracts/integrations/NFTfiV2/) - NFTfiV2 Note Adapter
    - [`ArcadeV1/`](contracts/integrations/ArcadeV1/) - ArcadeV1 Note Adapter
  - [`test/`](contracts/test/) - Testing contracts
    - [`lending/`](contracts/test/lending/) - Test lending platform
      - [`TestNoteToken.sol`](contracts/test/lending/TestNoteToken.sol) - Note token for TestLendingPlatform
      - [`TestLendingPlatform.sol`](contracts/test/lending/TestLendingPlatform.sol) - TestLendingPlatform implementation
      - [`TestNoteAdapter.sol`](contracts/test/lending/TestNoteAdapter.sol) - Note adapter for TestLendingPlatform
    - [`TestERC20.sol`](contracts/test/TestERC20.sol) - Test ERC20 token
    - [`TestERC721.sol`](contracts/test/TestERC721.sol) - Test ERC721 token
    - [`MockLoanPriceOracle.sol`](contracts/test/MockLoanPriceOracle.sol) - Mock LoanPriceOracle
    - [`TestLPTokenUpgrade.sol`](contracts/test/TestLPTokenUpgrade.sol) - Test LPToken upgrade
    - [`TestVaultUpgrade.sol`](contracts/test/TestVaultUpgrade.sol) - Test Vault upgrade
- [`test/`](test/) - Unit tests
  - [`TestLendingPlatform.spec.ts`](test/TestLendingPlatform.spec.ts) - TestLendingPlatform unit tests
  - [`LoanPriceOracle.spec.ts`](test/LoanPriceOracle.spec.ts) - LoanPriceOracle unit tests
  - [`LPToken.spec.ts`](test/LPToken.spec.ts) - LPToken unit tests
  - [`Vault.spec.ts`](test/Vault.spec.ts) - Vault unit tests
  - [`Vault.accounting.spec.ts`](test/Vault.accounting.spec.ts) - Vault accounting unit tests
  - [`Vault.keeper.spec.ts`](test/Vault.keeper.spec.ts) - Vault keeper integration unit tests
  - [`Integration.spec.ts`](test/Integration.spec.ts) - Integration test
  - [`helpers/`](test/helpers/) - Test helpers
    - [`EventUtilities.ts`](test/helpers/EventUtilities.ts) - Event helper functions
    - [`FixedPointHelpers.ts`](test/helpers/FixedPointHelpers.ts) - Fixed point math helper functions
    - [`RandomHelpers.ts`](test/helpers/RandomHelpers.ts) - Random number helper functions
    - [`LoanPriceOracleHelpers.ts`](test/helpers/LoanPriceOracleHelpers.ts) - Loan price oracle helper functions
    - [`VaultHelpers.ts`](test/helpers/VaultHelpers.ts) - Vault helper functions
- [`scripts/`](scripts/) - Scripts
  - [`deploy-simulation.ts`](scripts/deploy-simulation.ts) - Simulation deployment
  - [`deployment-manager.ts`](scripts/deployment-manager.ts) - Deployment manager
- [`deployments/`](deployments/) - Deployments
- [`docs/`](docs/) - Documentation
  - [`SECURITY.md`](docs/SECURITY.md) - Security notes
- [`hardhat.config.ts`](hardhat.config.ts) - Hardhat configuration
- [`tsconfig.json`](tsconfig.json) - TypeScript configuration
- [`package.json`](package.json) - npm package metadata
- [`package-lock.json`](package-lock.json) - npm package lock
- [`README.md`](README.md) - This README

## License

MetaStreet v1 Contracts are primary BUSL-1.1 [licensed](LICENSE). Interfaces are MIT [licensed](contracts/interfaces/LICENSE).


CHANGELOG.md
* Vault v1.5 - 02/27/2023
    * Add support for optional note seller approval.

* XY3NoteAdapter v1.0 - 02/23/2023
    * Initial release.

* ArcadeV2NoteAdapter v1.2 - 12/08/2022
    * Add `UnreportedCollateralInventory()` error for collateral bundles missing
      inventory reporting.

* NFTfiV2NoteAdapter v1.3 - 11/30/2022
    * Add support for `DIRECT_LOAN_FIXED_COLLECTION` loan type for loans made
      from collection offers.

* Vault v1.4 - 11/21/2022
    * Add `getLoanAssets()` API to INoteAdapter.
    * Add `priceNote()` getter API to IVault.
    * Add support for pricing collateral bundles.
* ArcadeV2NoteAdapter v1.1 - 11/21/2022
    * Implement new `getLoanAssets()` API in INoteAdapter.
    * Add support for collateral bundles.
* NFTfiV2NoteAdapter v1.2 - 11/21/2022
    * Implement new `getLoanAssets()` API in INoteAdapter.

* Vault v1.3 - 11/18/2022
    * Add default constructor to disable initialization of implementation
      contract.
* LPToken v1.1 - 11/18/2022
    * Add default constructor to disable initialization of implementation
      contract.

* LoanPriceOracle v1.2 - 10/31/2022
    * Add purchase price validation to `priceLoan()`.

* NFTfiV2NoteAdapter v1.1 - 09/24/2022
    * Improve support for loan contract upgrades by modifying getters to look
      up loan contract dynamically.
    * Update supported loan type key to new loan contract deployment.

* Vault v1.2 - 08/25/2022
    * Add additional `utilization()` getter to return utilization computed with
      additional loan balance.

* Vault v1.1 - 07/25/2022
    * Modify `sellNote()` and `sellNoteAndDeposit()` APIs to return executed
      purchase price.

* LoanPriceOracle v1.1 - 07/19/2022
    * Add `priceLoanRepayment()` API.

* LoanPriceOracle v1.0 - 05/31/2022
    * Initial release.
* Vault v1.0 - 05/31/2022
    * Initial release.


hardhat.config.ts
import * as dotenv from "dotenv";

import { HardhatUserConfig, task } from "hardhat/config";
import "@nomiclabs/hardhat-etherscan";
import "@nomiclabs/hardhat-waffle";
import '@openzeppelin/hardhat-upgrades';
import "@typechain/hardhat";
import "hardhat-gas-reporter";
import "solidity-coverage";

dotenv.config();

const config: HardhatUserConfig = {
  solidity: {
    compilers: [
      {
        version: "0.8.9",
        settings: {
          optimizer: {
            enabled: true,
            runs: 200,
          },
        },
      },
    ],
  },
  networks: {
    mainnet: {
      url: process.env.MAINNET_URL || "",
    },
    rinkeby: {
      url: process.env.RINKEBY_URL || "",
      accounts:
        process.env.PRIVATE_KEY !== undefined ? [process.env.PRIVATE_KEY] : [],
    },
    goerli: {
      url: process.env.GOERLI_URL || "",
      accounts:
        process.env.PRIVATE_KEY !== undefined ? [process.env.PRIVATE_KEY] : [],
    },
  },
  gasReporter: {
    enabled: process.env.REPORT_GAS !== undefined,
    currency: "USD",
  },
  etherscan: {
    apiKey: process.env.ETHERSCAN_API_KEY,
  },
  typechain: {
    outDir: 'typechain',
  },
  mocha: {
    timeout: 60000
  },
};

export default config;


subgraph/mapping/promissoryNotes.ts
import { Address, store } from "@graphprotocol/graph-ts";
import { Transfer } from "../generated/IERC721/IERC721";
import { PromissoryNote } from "../generated/schema";

import { getOrInitUser } from "./helpers";

const ZERO_ADDRESS = Address.zero();

export function handleTransfer(event: Transfer): void {
  const contractAddress = event.address;
  const tokenId = event.params.tokenId;
  const noteId = `${contractAddress.toHexString()}-${tokenId.toString()}`;

  // mint
  if (event.params.from.equals(ZERO_ADDRESS)) {
    const toUser = getOrInitUser(event.params.to);
    const promissoryNote = new PromissoryNote(noteId);
    promissoryNote.contractAddress = contractAddress;
    promissoryNote.tokenId = tokenId;
    promissoryNote.owner = toUser.id;
    promissoryNote.save();
  }
  // burn
  else if (event.params.to.equals(ZERO_ADDRESS)) {
    const promissoryNote = PromissoryNote.load(noteId);
    if (promissoryNote) {
      store.remove("PromissoryNote", noteId);
    }
  }
  // transfer
  else {
    const toUser = getOrInitUser(event.params.to);
    const promissoryNote = PromissoryNote.load(noteId);
    if (promissoryNote) {
      promissoryNote.owner = toUser.id;
      promissoryNote.save();
    }
  }
}


scripts/deploy-simulation.ts
import { ethers } from "hardhat";

import { BigNumberish } from "@ethersproject/bignumber";

import { TestLendingPlatform } from "../typechain";
import { extractEvent } from "../test/helpers/EventUtilities";
import { elapseTime } from "../test/helpers/VaultHelpers";
import { FixedPoint } from "../test/helpers/FixedPointHelpers";
import {
  UtilizationParameters,
  CollateralParameters,
  encodeUtilizationParameters,
  encodeCollateralParameters,
  computePiecewiseLinearModel,
} from "../test/helpers/LoanPriceOracleHelpers";

async function main() {
  const accounts = await ethers.getSigners();
  console.log("Deploying from account #9 (%s)\n", accounts[9].address);

  const TestERC20 = await ethers.getContractFactory("TestERC20", accounts[9]);
  const TestERC721 = await ethers.getContractFactory("TestERC721", accounts[9]);
  const TestLendingPlatformFactory = await ethers.getContractFactory("TestLendingPlatform", accounts[9]);
  const VaultRegistry = await ethers.getContractFactory("VaultRegistry", accounts[9]);
  const TestNoteAdapter = await ethers.getContractFactory("TestNoteAdapter", accounts[9]);
  const StaticCollateralOracle = await ethers.getContractFactory("StaticCollateralOracle", accounts[9]);
  const LoanPriceOracle = await ethers.getContractFactory("LoanPriceOracle", accounts[9]);
  const LPToken = await ethers.getContractFactory("LPToken", accounts[9]);
  const Vault = await ethers.getContractFactory("Vault", accounts[9]);

  /* Deploy DAI */
  const daiTokenContract = await TestERC20.deploy("DAI", "DAI", 18, ethers.utils.parseEther("1000000"));
  await daiTokenContract.deployed();
  console.log("DAI Token Contract:     ", daiTokenContract.address);

  /* Deploy WETH */
  const wethTokenContract = await TestERC20.deploy("WETH", "WETH", 18, ethers.utils.parseEther("1000000"));
  await wethTokenContract.deployed();
  console.log("WETH Token Contract:    ", wethTokenContract.address);

  /* Deploy BAYC */
  const baycTokenContract = await TestERC721.deploy(
    "BoredApeYachtClub",
    "BAYC",
    "ipfs://QmeSjSinHpPnmXmspMjwiXyN6zS4E9zccariGR3jxcaWtq/"
  );
  await baycTokenContract.deployed();
  console.log("BAYC Token Contract:    ", baycTokenContract.address);

  console.log("");

  /* Deploy DAI Test Lending Platform */
  const daiTestLendingPlatform = await TestLendingPlatformFactory.deploy(daiTokenContract.address);
  await daiTestLendingPlatform.deployed();
  console.log("DAI Lending Platform:   ", daiTestLendingPlatform.address);
  console.log("       Note Token Address: ", await daiTestLendingPlatform.noteToken());
  console.log("");

  /* Deploy WETH Test Lending Platform */
  const wethTestLendingPlatform = await TestLendingPlatformFactory.deploy(wethTokenContract.address);
  await wethTestLendingPlatform.deployed();
  console.log("WETH Lending Platform:  ", wethTestLendingPlatform.address);
  console.log("       Note Token Address: ", await wethTestLendingPlatform.noteToken());
  console.log("");

  /* Deploy Vault Registry */
  const vaultRegistry = await VaultRegistry.deploy();
  await vaultRegistry.deployed();
  console.log("Vault Registry:         ", vaultRegistry.address);
  console.log("");

  /* Deploy DAI Test Note Adapter */
  const daiTestNoteAdapter = await TestNoteAdapter.deploy(daiTestLendingPlatform.address);
  await daiTestNoteAdapter.deployed();
  console.log("DAI Test Note Adapter:  ", daiTestNoteAdapter.address);

  /* Deploy WETH Test Note Adapter */
  const wethTestNoteAdapter = await TestNoteAdapter.deploy(wethTestLendingPlatform.address);
  await wethTestNoteAdapter.deployed();
  console.log("WETH Test Note Adapter: ", wethTestNoteAdapter.address);

  /* Deploy Static Collateral Oracle for DAI */
  const daiStaticCollateralOracle = await StaticCollateralOracle.deploy(daiTokenContract.address);
  await daiStaticCollateralOracle.deployed();
  console.log("DAI Collateral Oracle:  ", daiStaticCollateralOracle.address);

  /* Deploy Static Collateral Oracle for WETH */
  const wethStaticCollateralOracle = await StaticCollateralOracle.deploy(wethTokenContract.address);
  await wethStaticCollateralOracle.deployed();
  console.log("WETH Collateral Oracle: ", wethStaticCollateralOracle.address);

  /* Deploy Loan Price Oracle for DAI */
  const daiLoanPriceOracle = await LoanPriceOracle.deploy(daiStaticCollateralOracle.address);
  await daiLoanPriceOracle.deployed();
  console.log("DAI Loan Price Oracle:  ", daiLoanPriceOracle.address);

  /* Deploy Loan Price Oracle for WETH */
  const wethLoanPriceOracle = await LoanPriceOracle.deploy(wethStaticCollateralOracle.address);
  await wethLoanPriceOracle.deployed();
  console.log("WETH Loan Price Oracle: ", wethLoanPriceOracle.address);

  console.log("");

  /* Deploy DAI Vault Senior LP token */
  const daiBlueChipVaultSeniorLPToken = await LPToken.deploy();
  await daiBlueChipVaultSeniorLPToken.deployed();
  await daiBlueChipVaultSeniorLPToken.initialize("Senior LP Token", "msLP-BC-DAI");

  /* Deploy DAI Vault Junior LP token */
  const daiBlueChipVaultJuniorLPToken = await LPToken.deploy();
  await daiBlueChipVaultJuniorLPToken.deployed();
  await daiBlueChipVaultJuniorLPToken.initialize("Junior LP Token", "mjLP-BC-DAI");

  /* Deploy DAI Vault */
  const daiBlueChipVault = await Vault.deploy();
  await daiBlueChipVault.deployed();
  await daiBlueChipVault.initialize(
    "Blue Chip / DAI",
    daiTokenContract.address,
    daiLoanPriceOracle.address,
    daiBlueChipVaultSeniorLPToken.address,
    daiBlueChipVaultJuniorLPToken.address
  );
  await daiBlueChipVault.deployed();

  /* Transfer ownership of LP tokens to DAI Vault */
  await daiBlueChipVaultSeniorLPToken.transferOwnership(daiBlueChipVault.address);
  await daiBlueChipVaultJuniorLPToken.transferOwnership(daiBlueChipVault.address);

  console.log("Blue Chip DAI Vault:    ", daiBlueChipVault.address);
  console.log("               Vault Name: ", await daiBlueChipVault.name());
  console.log(
    "   Senior LP Token Symbol: ",
    await (await ethers.getContractAt("IERC20Metadata", daiBlueChipVaultSeniorLPToken.address)).symbol()
  );
  console.log("  Senior LP Token Address: ", daiBlueChipVaultSeniorLPToken.address);
  console.log(
    "   Junior LP Token Symbol: ",
    await (await ethers.getContractAt("IERC20Metadata", daiBlueChipVaultJuniorLPToken.address)).symbol()
  );
  console.log("  Junior LP Token Address: ", daiBlueChipVaultSeniorLPToken.address);
  console.log("");

  /* Deploy WETH Vault Senior LP token */
  const wethBlueChipVaultSeniorLPToken = await LPToken.deploy();
  await wethBlueChipVaultSeniorLPToken.deployed();
  await wethBlueChipVaultSeniorLPToken.initialize("Senior LP Token", "msLP-BC-WETH");

  /* Deploy WETH Vault Junior LP token */
  const wethBlueChipVaultJuniorLPToken = await LPToken.deploy();
  await wethBlueChipVaultJuniorLPToken.deployed();
  await wethBlueChipVaultJuniorLPToken.initialize("Junior LP Token", "mjLP-BC-WETH");

  /* Deploy WETH Vault */
  const wethBlueChipVault = await Vault.deploy();
  await wethBlueChipVault.deployed();
  await wethBlueChipVault.initialize(
    "Blue Chip / WETH",
    wethTokenContract.address,
    wethLoanPriceOracle.address,
    wethBlueChipVaultSeniorLPToken.address,
    wethBlueChipVaultJuniorLPToken.address
  );
  await wethBlueChipVault.deployed();

  /* Transfer ownership of LP tokens to WETH Vault */
  await wethBlueChipVaultSeniorLPToken.transferOwnership(wethBlueChipVault.address);
  await wethBlueChipVaultJuniorLPToken.transferOwnership(wethBlueChipVault.address);

  console.log("Blue Chip WETH Vault:    ", wethBlueChipVault.address);
  console.log("               Vault Name: ", await wethBlueChipVault.name());
  console.log(
    "   Senior LP Token Symbol: ",
    await (await ethers.getContractAt("IERC20Metadata", wethBlueChipVaultSeniorLPToken.address)).symbol()
  );
  console.log("  Senior LP Token Address: ", wethBlueChipVaultSeniorLPToken.address);
  console.log(
    "   Junior LP Token Symbol: ",
    await (await ethers.getContractAt("IERC20Metadata", wethBlueChipVaultJuniorLPToken.address)).symbol()
  );
  console.log("  Junior LP Token Address: ", wethBlueChipVaultJuniorLPToken.address);
  console.log("");

  await daiBlueChipVault.setNoteAdapter(await daiTestLendingPlatform.noteToken(), daiTestNoteAdapter.address);
  console.log("Attached DAI Test Note Adapter to Blue Chip DAI Vault");

  await wethBlueChipVault.setNoteAdapter(await wethTestLendingPlatform.noteToken(), wethTestNoteAdapter.address);
  console.log("Attached WETH Test Note Adapter to Blue Chip WETH Vault");
  console.log("");

  await vaultRegistry.registerVault(daiBlueChipVault.address);
  console.log("Registered Blue Chip DAI Vault with Vault Registry");

  await vaultRegistry.registerVault(wethBlueChipVault.address);
  console.log("Registered Blue Chip WETH Vault with Vault Registry");
  console.log("");

  console.log("Lender is      account #0 (%s)", accounts[0].address);
  console.log("Borrower is    account #1 (%s)", accounts[1].address);
  console.log("Depositer 1 is account #2 (%s)", accounts[2].address);
  console.log("Depositer 2 is account #3 (%s)", accounts[3].address);
  console.log("");

  await daiTokenContract.transfer(accounts[0].address, ethers.utils.parseEther("1000"));
  await daiTokenContract.transfer(accounts[1].address, ethers.utils.parseEther("1000"));
  await daiTokenContract.transfer(accounts[2].address, ethers.utils.parseEther("1000"));
  await daiTokenContract.transfer(accounts[3].address, ethers.utils.parseEther("1000"));
  console.log("Transferred 1000 DAI to account #0, #1, #2, #3");

  await wethTokenContract.transfer(accounts[0].address, ethers.utils.parseEther("1000"));
  await wethTokenContract.transfer(accounts[1].address, ethers.utils.parseEther("1000"));
  await wethTokenContract.transfer(accounts[2].address, ethers.utils.parseEther("1000"));
  await wethTokenContract.transfer(accounts[3].address, ethers.utils.parseEther("1000"));
  console.log("Transferred 1000 WETH to account #0, #1, #2, #3");

  await baycTokenContract.mint(accounts[1].address, 123);
  await baycTokenContract.mint(accounts[1].address, 456);
  await baycTokenContract.mint(accounts[1].address, 768);
  console.log("Minted BAYC #123, #456, #768 to account #1");

  await baycTokenContract.connect(accounts[1]).setApprovalForAll(daiTestLendingPlatform.address, true);
  console.log("Approved BAYC transfer for DAI Test Lending Platform for account #1");

  await baycTokenContract.connect(accounts[1]).setApprovalForAll(wethTestLendingPlatform.address, true);
  console.log("Approved BAYC transfer for WETH Test Lending Platform for account #1");

  await daiTokenContract.connect(accounts[1]).approve(daiTestLendingPlatform.address, ethers.constants.MaxUint256);
  console.log("Approved DAI transfer for DAI Test Lending Platform for account #1");

  await wethTokenContract.connect(accounts[1]).approve(wethTestLendingPlatform.address, ethers.constants.MaxUint256);
  console.log("Approved WETH transfer for WETH Test Lending Platform for account #1");

  await daiTokenContract.connect(accounts[0]).approve(daiTestLendingPlatform.address, ethers.constants.MaxUint256);
  console.log("Approved DAI transfer for DAI Test Lending Platform for account #0");

  await wethTokenContract.connect(accounts[0]).approve(wethTestLendingPlatform.address, ethers.constants.MaxUint256);
  console.log("Approved DAI transfer for WETH Test Lending Platform for account #0");

  console.log("");

  /* Setup parameters for loan price oracles */
  const utilizationParameters: UtilizationParameters = computePiecewiseLinearModel({
    minRate: FixedPoint.normalizeRate("0.05"),
    targetRate: FixedPoint.normalizeRate("0.10"),
    maxRate: FixedPoint.normalizeRate("2.00"),
    target: FixedPoint.from("0.90"),
    max: FixedPoint.from("1.00"),
  });

  const collateralValue = ethers.utils.parseEther("100");

  const collateralParameters: CollateralParameters = {
    active: true,
    loanToValueRateComponent: computePiecewiseLinearModel({
      minRate: FixedPoint.normalizeRate("0.05"),
      targetRate: FixedPoint.normalizeRate("0.10"),
      maxRate: FixedPoint.normalizeRate("2.00"),
      target: FixedPoint.from("0.30"),
      max: FixedPoint.from("0.60"),
    }),
    durationRateComponent: computePiecewiseLinearModel({
      minRate: FixedPoint.normalizeRate("0.05"),
      targetRate: FixedPoint.normalizeRate("0.10"),
      maxRate: FixedPoint.normalizeRate("2.00"),
      target: ethers.BigNumber.from(30 * 86400).mul(ethers.constants.WeiPerEther),
      max: ethers.BigNumber.from(90 * 86400).mul(ethers.constants.WeiPerEther),
    }),
    rateComponentWeights: [5000, 2500, 2500],
  };

  await daiStaticCollateralOracle.setCollateralValue(baycTokenContract.address, collateralValue);
  console.log("Setup collateral value for DAI Collateral Oracle");

  await daiLoanPriceOracle.setUtilizationParameters(encodeUtilizationParameters(utilizationParameters));
  console.log("Setup utilization parameters for DAI Loan Price Oracle");

  await daiLoanPriceOracle.setCollateralParameters(
    baycTokenContract.address,
    encodeCollateralParameters(collateralParameters)
  );
  console.log("Setup BAYC collateral parameters for DAI Loan Price Oracle");

  await wethStaticCollateralOracle.setCollateralValue(baycTokenContract.address, collateralValue);
  console.log("Setup collateral value for WETH Collateral Oracle");

  await wethLoanPriceOracle.setUtilizationParameters(encodeUtilizationParameters(utilizationParameters));
  console.log("Setup utilization parameters for WETH Loan Price Oracle");

  await wethLoanPriceOracle.setCollateralParameters(
    baycTokenContract.address,
    encodeCollateralParameters(collateralParameters)
  );
  console.log("Setup BAYC collateral parameters for WETH Loan Price Oracle");

  console.log("");

  async function createLoan(
    lendingPlatform: TestLendingPlatform,
    collateralTokenAddress: string,
    collateralTokenId: number,
    principal: BigNumberish,
    repayment: BigNumberish,
    duration: number
  ): Promise<BigNumberish> {
    const lendTx = await lendingPlatform.lend(
      accounts[1].address,
      accounts[0].address,
      collateralTokenAddress,
      collateralTokenId,
      principal,
      repayment,
      duration
    );
    return (await extractEvent(lendTx, lendingPlatform, "LoanCreated")).args.loanId;
  }

  let loanId: BigNumberish;

  loanId = await createLoan(
    daiTestLendingPlatform,
    baycTokenContract.address,
    123,
    ethers.utils.parseEther("30"),
    ethers.utils.parseEther("33"),
    30 * 86400
  );
  console.log("Created DAI Loan ID %s:  Borrower account #1, Lender account #0,", loanId);
  console.log("                        Principal 10 DAI, Repayment 10.42 DAI, Duration 30 days,");
  console.log("                        Collateral Token ID 123\n");

  /* Fast-forward time by 15 days */
  await elapseTime(15 * 86400);
  console.log("Fast-forwarded time by 15 days\n");

  loanId = await createLoan(
    wethTestLendingPlatform,
    baycTokenContract.address,
    456,
    ethers.utils.parseEther("20"),
    ethers.utils.parseEther("21"),
    30 * 86400
  );
  console.log("Created WETH Loan ID %s: Borrower account #1, Lender account #0,", loanId);
  console.log("                        Principal 30 WETH, Repayment 30.12 WETH, Duration 30 days,");
  console.log("                        Collateral Token ID 456\n");

  loanId = await createLoan(
    wethTestLendingPlatform,
    baycTokenContract.address,
    768,
    ethers.utils.parseEther("15"),
    ethers.utils.parseEther("16"),
    60 * 86400
  );
  console.log("Created WETH Loan ID %s: Borrower account #1, Lender account #0,", loanId);
  console.log("                        Principal 15 WETH, Repayment 15.85 WETH, Duration 60 days,");
  console.log("                        Collateral Token ID 768");
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
